src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DocumentDesigner.cs

```diff
@@ -877,69 +876,57 @@ private void OnUndoing(object source, EventArgs e)
     /// </summary>
     private void OnComponentAdded(object source, ComponentEventArgs ce)
     {
-        IDesignerHost host = (IDesignerHost)GetService(typeof(IDesignerHost));
-        if (host is not null)
+        if (!TryGetService(out IDesignerHost host))
         {
-            IComponent component = ce.Component;
+            return;
+        }
 
-            bool addControl = true;
+        IComponent component = ce.Component;
 
-            // This is the mirror to logic in ParentControlDesigner. The component should be
-            // added somewhere, and this logic decides where.
+        // This is the mirror to logic in ParentControlDesigner. The component should be
+        // added somewhere, and this logic decides where.
 
-            // LETS SEE IF WE ARE TOOLSTRIP in which case we want to get added
-            // to the componentTray even though this is a control..
-            // We should think of implementing an interface so that we can have many more
-            // controls behaving like this.
+        // If the component is a toolstrip or a top level form, we should add to the tray.
 
-            if (host.GetDesigner(component) is not ToolStripDesigner td)
-            {
-                ControlDesigner cd = host.GetDesigner(component) as ControlDesigner;
-                if (cd is not null)
-                {
-                    if (cd.Control is not Form form || !form.TopLevel)
-                    {
-                        addControl = false;
-                    }
-                }
-            }
+        IDesigner designer = host.GetDesigner(component);
+        bool addControl = designer is ToolStripDesigner
+            || (designer is ControlDesigner cd && cd.Control is Form form && form.TopLevel);
```

Comment:
I think this needs to be
```suggestion
        bool addControl = designer is ToolStripDesigner
            || designer is not ControlDesigner cd
            || (cd.Control is Form form && form.TopLevel);
```